### ISSUE TITLE
Change runhistory for parallel usage

### ISSUE THREAD

**Issue by SgnJp** [automl/SMAC3#96]
A few changes need to be done to runhistory to be possible to use it in parallel:
1.  Locks on internal data are needed
2. One more status of run should be added ("QUEUED"), for runs which are currently running or just in the queue to run. And corresponding method to get access for this runs.
3. get_runs_for_config is too slow for a large number of runs. To make it faster I propose to add a dict from config_id to dict from (instance, seed) to RunValue. So runhistory will consume more space, but the method will run in logarithmic time instead of linear.

This changes are already done in my [SMAC fork](https://github.com/SgnJp/SMAC3/blob/master/smac/runhistory/runhistory.py).

You can also see the diff between forks [here](https://github.com/automl/SMAC3/compare/master...SgnJp:master), you are interested in file "runhistory.py"

I need to know your opinion, whether it is ok, or it may cause some errors at some cases, which I didn't take into account

**Comment by mfeurer** [comment]
Hey,

That all makes sense and I don't see any compability issues right now. Some things you might consider:

1. I think that also the load and save methods of the runhistory must be encapsulated in a mutex.
2. use `is config_id is not None` instead of `not config_id is None`
3. How much overhead did the current implementation of `get_runs_for_config` produce? How much does this overhead decrease?

**Comment by SgnJp** [comment]
1-2. Agree
3. I don't know how to quantify overhead. Let's say we have n configuration, with k runs each. Right now `get_runs_for_config` is looping throw all nk data points. I propose to change this loop in favor of extraction from dict, and one loop, but throw k data points. So it should run n times faster. (Btw, it is not logarithmic as I stated before).

Before I had a problem that my workers were executing new runs, and master node(IntensifierParallel) was not able to process all these information, and  `get_runs_for_config` was a bottleneck. With new representation I don't have this problem.

**Comment by mfeurer** [comment]
This sounds reasonable. To show the difference, you could use the python profiler on both versions of the code and post the difference (although this would not be necessary for me to accept such a change).

**Comment by mfeurer** [comment]
Due to an unrelated performance issue which made me profile SMAC3, I can now confirm that `get_runs_for_config` can take up to 80% of the time needed to add a new run to the runhistory.

**Comment by mlindauer** [comment]
I think we can close this PR. We added to the runhistory a dictionary `_configid_to_inst_seed` with which a faster access is possible. For the time being, we will not change the runhistory to support the parallel version. A parallel runhistory should inherit from this runhistory and make the required changes.
